### DISCUSSION TITLE
How can I access registers directly?

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#2310] [Q&A] [answered]
I am trying to access HW registers directly, for example Timer 2

If I write
```
  TIM2->CCMR1 = 1;
  Serial.print("TIM2->CCMR1: ");
  Serial.println(TIM2->CCMR1);
  TIM2->CR1 = 1;
  Serial.print("TIM2->CR1: ");
  Serial.println(TIM2->CR1);
```
I get

> TIM2->CCMR1: 0
> TIM2->CR1: 0

but, when I use API functions first, like
```
  HardwareTimer *MyTim = new HardwareTimer(TIM2);
  MyTim->setMode(1, TIMER_INPUT_FREQ_DUTY_MEASUREMENT, PA5);
  Serial.print("TIM2->CCMR1: ");
  Serial.println(TIM2->CCMR1);
  Serial.print("TIM2->CR1: ");
  Serial.println(TIM2->CR1);
```
I get

> TIM2->CCMR1: 513
> TIM2->CR1: 128

Now I also can assign custom values to the registers directly...
It is even sufficient to cerate the HardwareTimer object

Do I have to "activate" the TIM2 register variable somehow?

**Comment by fpistm** [comment, marked as answer]
Hi @MasteringTheMess 
As a first guess, I would recommend to enable the clock

`__HAL_RCC_TIM2_CLK_ENABLE();`

**Comment by PerennialNovice** [comment]
A very good guess, indeed :)

That did the trick, thank you.

can also be achieved by configuring the register directly:

`RCC->APB1ENR1 |= RCC_APB1ENR1_TIM2EN;`

**Reply by fpistm** [reply]
Welcome

**Comment by PerennialNovice** [comment]
the corresponding information is also in the datasheet, but not at the peripheral register description.
It is included at the peripheral clock enable register !
